internal/statesync/statesync.go

```diff
@@ -0,0 +1,316 @@
+// receive and validate snapshots from the network and initialize the database state.
+// It takes the database configuration, snapshot directory, and the trusted snapshot providers.
+// Trusted snapshot providers are special nodes in the network trusted by the nodes and
+// have snashopt creation enabled. These nodes are responsible for creating and validating snapshots.
+func NewStateSyncer(ctx context.Context, cfg *DBConfig, snapshotDir string, providers []string, logger log.Logger) *StateSyncer {
+
+	ss := &StateSyncer{
+		dbConfig:     cfg,
+		snapshotsDir: snapshotDir,
+		log:          logger,
+	}
+
+	for _, s := range providers {
+		clt, err := ChainRPCClient(s)
+		if err != nil {
+			logger.Error("Failed to create rpc client", log.String("server", s), log.Error(err))
+			return nil
+		}
+		ss.snapshotProviders = append(ss.snapshotProviders, clt)
+	}
+
+	// Ensure that the snasphot directory exists and is empty
+	if err := os.RemoveAll(snapshotDir); err != nil {
+		logger.Error("Failed to delete snapshot directory", log.String("dir", snapshotDir), log.Error(err))
+		return nil
+	}
+	if err := os.MkdirAll(snapshotDir, 0755); err != nil {
+		logger.Error("Failed to create snapshot directory", log.String("dir", snapshotDir), log.Error(err))
+		return nil
+	}
+
+	return ss
+}
+
+// OfferSnapshot checks if the snapshot is valid and kicks off the state sync process
+// accepted snapshot is stored on disk for later processing
+func (ss *StateSyncer) OfferSnapshot(snapshot *Snapshot) error {
+	ss.log.Info("Offering snapshot", log.Int("height", int64(snapshot.Height)), log.Uint("format", snapshot.Format), log.String("App Hash", fmt.Sprintf("%x", snapshot.SnapshotHash)))
+
+	// Check if we are already in the middle of a snapshot
+	if ss.inProgress {
+		return ErrStateSyncInProgress
+	}
+
+	// Validate the snapshot
+	err := ss.validateSnapshot(*snapshot)
+	if err != nil {
+		return err
+	}
+
+	ss.snapshot = snapshot
+	ss.inProgress = true
+	ss.chunks = make(map[uint32]bool, snapshot.ChunkCount)
+	ss.rcvdChunks = 0
+	return nil
+}
+
+// ApplySnapshotChunk accepts a chunk and stores it on disk for later processing if valid
+// If all chunks are received, it starts the process of restoring the database
+func (ss *StateSyncer) ApplySnapshotChunk(ctx context.Context, chunk []byte, index uint32) (bool, bool, error) {
+	if !ss.inProgress {
+		return false, false, ErrStateSyncNotInProgress
+	}
+
+	// Check if the chunk has already been applied
+	applied, ok := ss.chunks[index]
+	if ok && applied {
+		return false, false, nil
+	}
+
+	// Check if the chunk index is valid
+	if index >= ss.snapshot.ChunkCount {
+		ss.log.Error("Invalid chunk index", log.Uint("index", index), log.Uint("chunk-count", ss.snapshot.ChunkCount))
+		return false, false, ErrRejectSnapshotChunk
+	}
+
+	// Validate the chunk hash
+	chunkHash := sha256.Sum256(chunk)
+	if !bytes.Equal(chunkHash[:], ss.snapshot.ChunkHashes[index][:]) {
```

Comment:
Arrays are `comparable`.  This was one of the benefits of converting the  hashes to arrays in the last iteration.  Just `if chunkHash != ss.snapshot.ChunkHashes[index] {` works.